handbook.md

```diff
@@ -0,0 +1,233 @@
+* Set ```:mark_translated: true``` in ```config/settings.yaml``` to spot missing string i18n extractions.
+* Concerns and new classes are added to app/, not lib/.
+* Favor ```.present?``` over ```.nil?```. Only use the latter if you are truly checking for nil, which is uncommon.
+* If adding new model attributes, use scoped search definitions where appropriate.
+* Virtual fields must have the option ```:only_explicit``` when added to scoped search.
+* Only catch exceptions that are expected. Wrap them in `Foreman::WrappedException`
+* When you need to catch a `Foreman::Exception` make sure the ERF code is displayed to the user. Send a pull request to [theforeman.org](https://github.com/theforeman.org) with possible fixes for the ERF code.
+
+####Don't
+* Use ```.to_sym```, ```.send```, ```eval``` or other reflection on untrusted inputs.
+* Catch unexpected exceptions, or `Foreman::Exception`.
+* Use single character variable names or abbreviations. The keyboard erosion you might save by doing that is not worth it.
+
+### Tests
+We use MiniTest::Unit syntax.
+
+* Use ```test 'description' do``` instead of ```def test_description```
+* Use custom assertions for variable status checks, e.g: ```assert_empty variable``` instead of ```assert variable.empty?```. These assertions produce errors that are easier to read.
+* Use ```:success, :not_authorized, etc...```, instead of actual HTTP status codes.
+* Favor FactoryGirl over fixtures. Only use fixtures when the object is created very often throughout the test suite.
+* If your tests are repeating code except for a couple of changes, wrap it in a `context` block, or write a helper that takes in the changes through arguments and generates the tests.
+* Use stubs and mocks extensively to test external calls, slow calls not related with the actual test, and also to avoid testing framework features. They make our test base much faster.
+* Use `setup_user` if you need to create an user with special permissions and roles instead of doing it yourself in the test.
+* For functional tests, if several controllers share the same behavior, extract it to a [shared functional concern](https://github.com/theforeman/foreman/blob/develop/test/functional/shared/report_host_permissions_test.rb)
+
+
+### Strings and translations
+
+Be mindful of our [Translating](http://projects.theforeman.org/projects/foreman/wiki/Translating) guide. A quick overview:
+
+#### Ruby
+
+*    To translate a string use `_('My string')`
+*    To translate string with a parameter use `_("String with param: %s") % param`
+*    To translate string with more than one parameters do not use `_("Params: %s and %s") % [param1, param2]` which is translator-unfriendly
+*    Therefore for more than one parameters use `_("Params: %{a} and %{b}") % {:a => foo, :b => bar}`
+*    Note that parameters are not translated. Variables should not contain text that needs translation.
+*    To mark something for translation (but not translate) use `N_("String")`
```

Comment:
add: "Ensure you later translate the text before it's displayed with `_(variable)`.`